### DISCUSSION TITLE
ccache with msbuild works, but building the clean target clears the cache

### DISCUSSION THREAD

**Discussion by craigscott-crascit** [ccache/ccache#1140] [General]
I have been experimenting with getting ccache working with CMake's Visual Studio generators. With ccache 4.6.1, I have it working when building inside the Visual Studio IDE, and when building via the command line (msbuild). I'll eventually document the steps, hopefully with an updated blog article (old one [here](https://crascit.com/2016/04/09/using-ccache-with-cmake/), needs updating and simplifying now).

There's one not-so-nice behavior I've encountered though which I don't know if it is possible to prevent. When you do a build, it appears that all files that get created or modified are recorded by the build tool. If you later build the `clean` target, it appears that the build tool revisits that list of files and removes all of them. This includes all the files that `ccache` stores in its cache, so a `clean` ends up also cleaning all the objects out of `ccache`'s cache too. This is undesirable, but I don't know how to prevent it. This occurs both when building inside the Visual Studio IDE and when using `msbuild` on the command line.

I'm curious if anyone has any thoughts on how we might be able to avoid the `clean` target from removing things from `ccache`'s cache? Or is there a way we can prevent it from recording `ccache`'s private cache files in the first place?

@cristianadam and others, thanks for your great work in making this possible, btw!

**Comment by craigscott-crascit** [comment]
For extra info, the guidance from the [buildcache docs](https://github.com/mbitsnbites/buildcache/blob/master/doc/usage.md#using-with-visual-studio--msbuild) suggests putting the cache under somewhere like `C:\ProgramData\Ccache` might be needed for MSBuild's FileTracker. My experiments indicate that doesn't help though, and neither does putting it under any of the [other paths](https://github.com/dotnet/msbuild/blob/9eb5d09e6cd262375e37a15a779d56ab274167c8/src/Utilities/TrackedDependencies/FileTracker.cs#L208) that it seems like the FileTracker should be ignoring. I still see these paths showing up in the .tlog files.

**Comment by cristianadam** [comment]
I just tried the following `CMakeLists.txt` with ccache 4.6.2 (got the tip from [CMake issue 16176](https://gitlab.kitware.com/cmake/cmake/-/issues/16176)):

```cmake
cmake_minimum_required(VERSION 3.15)

project(hello-ccache)

set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

find_program(ccache_exe ccache)
if (ccache_exe)
  file(COPY_FILE
    ${ccache_exe} ${CMAKE_BINARY_DIR}/cl.exe
    ONLY_IF_DIFFERENT)

  set(CMAKE_VS_GLOBALS
    "CLToolExe=cl.exe"
    "CLToolPath=${CMAKE_BINARY_DIR}"
    "TrackFileAccess=false"
    "UseMultiToolTask=true"
    "DebugInformationFormat=OldStyle"
  )

endif()

add_executable(hello-ccache main.cpp)

if (ccache_exe)
  add_custom_command(TARGET hello-ccache
    PRE_BUILD COMMAND ${ccache_exe} -z)
  add_custom_command(TARGET hello-ccache
    POST_BUILD COMMAND ${ccache_exe} -s)
endif()
```

Then ran the following commands, and everything worked as expected:

```
c:\Projects\c++\hello-ccache
$ cmake -G "Visual Studio 17 2022" -A x64 -S . -B build
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
-- The C compiler identification is MSVC 19.33.31629.0
-- The CXX compiler identification is MSVC 19.33.31629.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Projects/c++/hello-ccache/build

c:\Projects\c++\hello-ccache
$ cmake --build build --config Release
MSBuild version 17.3.0+f67e3d35e for .NET Framework
  Checking Build System
  Statistics zeroed
  Building Custom Rule C:/Projects/c++/hello-ccache/CMakeLists.txt
  main.cpp
  hello-ccache.vcxproj -> C:\Projects\c++\hello-ccache\build\Release\hello-ccache.exe
  Summary:
    Hits:               0 /     1 (0.00 %)
      Direct:           0 /     1 (0.00 %)
      Preprocessed:     0 /     1 (0.00 %)
    Misses:             1
      Direct:           1
      Preprocessed:     1
  Primary storage:
    Hits:               0 /     2 (0.00 %)
    Misses:             2
    Cache size (GB): 0.00 / 50.00 (0.00 %)

  Use the -v/--verbose option for more details.
  Building Custom Rule C:/Projects/c++/hello-ccache/CMakeLists.txt

c:\Projects\c++\hello-ccache
$ cmake --build build --config Release --target clean
MSBuild version 17.3.0+f67e3d35e for .NET Framework

c:\Projects\c++\hello-ccache
$ cmake --build build --config Release
MSBuild version 17.3.0+f67e3d35e for .NET Framework
  Checking Build System
  Statistics zeroed
  Building Custom Rule C:/Projects/c++/hello-ccache/CMakeLists.txt
  main.cpp
  hello-ccache.vcxproj -> C:\Projects\c++\hello-ccache\build\Release\hello-ccache.exe
  Summary:
    Hits:               1 /     1 (100.0 %)
      Direct:           1 /     1 (100.0 %)
      Preprocessed:     0 /     0
    Misses:             0
      Direct:           0
      Preprocessed:     0
  Primary storage:
    Hits:               2 /     2 (100.0 %)
    Misses:             0
    Cache size (GB): 0.00 / 50.00 (0.00 %)

  Use the -v/--verbose option for more details.
  Building Custom Rule C:/Projects/c++/hello-ccache/CMakeLists.txt

c:\Projects\c++\hello-ccache
$ cmake -E rm -rf build

c:\Projects\c++\hello-ccache
$ cmake -G "Visual Studio 17 2022" -A x64 -S . -B build
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
-- The C compiler identification is MSVC 19.33.31629.0
-- The CXX compiler identification is MSVC 19.33.31629.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Projects/c++/hello-ccache/build

c:\Projects\c++\hello-ccache
$ cmake --build build --config Release
MSBuild version 17.3.0+f67e3d35e for .NET Framework
  Checking Build System
  Statistics zeroed
  Building Custom Rule C:/Projects/c++/hello-ccache/CMakeLists.txt
  main.cpp
  hello-ccache.vcxproj -> C:\Projects\c++\hello-ccache\build\Release\hello-ccache.exe
  Summary:
    Hits:               1 /     1 (100.0 %)
      Direct:           1 /     1 (100.0 %)
      Preprocessed:     0 /     0
    Misses:             0
      Direct:           0
      Preprocessed:     0
  Primary storage:
    Hits:               2 /     2 (100.0 %)
    Misses:             0
    Cache size (GB): 0.00 / 50.00 (0.00 %)

  Use the -v/--verbose option for more details.
  Building Custom Rule C:/Projects/c++/hello-ccache/CMakeLists.txt
```

**Reply by craigscott-crascit** [reply]
Awesome, thanks! Looks like the `TrackFileAccess=false` is what I was missing. Much appreciated.

**Reply by stilllman** [reply]
I've been trying to integrate `ccache` into Visual Studio lately but still fails to have a setup that works correctly.

Setting `TrackFileAccess=false` entirely disable incremental builds on my side, meaning that files don't get recompiled when I modify them. If I set it back to `true`, then in addition to the `clean` issue mentioned in this ticket, I also have incorrect incremental build behavior, presumably because `msbuild` tracks the `ccache` files as both inputs and outputs (so everything is dirty all the time). Putting the `ccache` files in `%USERPROFILE%/AppData/Roaming` doesn't seem to mitigate the problem.

Did you find a way to prevent `msbuild` from messing with the `ccache` files?

**Reply by cristianadam** [reply]
It might be useful to ask the question at https://developercommunity.visualstudio.com/cpp or switch to using ninja as a build system with Visual Studio.

**Reply by craigscott-crascit** [reply]
@stilllman I don't see the behavior you describe. In a small test case, I have `TrackFileAccess=false` and if I modify one of the sources, it does get recompiled.

**Reply by stilllman** [reply]
Hmm ok that's good to know, I'll investigate more with a simpler project then (I'll try to add updates here if I find anything interesting).

**Reply by mohsin-TW** [reply]
@craigscott-crascit @cristianadam Does ccache + MSVC compiler work with precompiled headers? I tried but it did not work for me.

Your help would be greatly appreciated

**Comment by mohsin-TW** [comment]
@craigscott-crascit  @cristianadam Does ccache + MSVC compiler work with precompiled headers? I tried but it did not work for me.

Your help would be greatly appreciated.

**Comment by ftrofin** [comment]
I'm having the same issue. Using ccache 4.10.2 and MSVS 17.12.5 (the latter probably doesn't matter). Here is scoop:
1. To be able to compile with ccache & msbuild you need to set `"ClCompile.ObjectFileName" :  "$$(IntDir)%(Filename).obj",` (basically disable batch compilation). This will kill your build performance. To bring it back you need the MTT (MultiToolTask) see MTT section in this doc https://devblogs.microsoft.com/cppblog/cpp-build-throughput-investigation-and-tune-up/. You basically want the Directory.Build.props file in the root of your workspace. (I don't use the MSBuild Resource Manager, just the MTT).
2. When you build via your ccache wrapper, VS will keep track of what files are read and written during the build, so it will also record all the ccache writes to the ccache folder into a tlog file. That tlog file is then used (I presume) when you invoke the "Clean" action and your files in the ccache cache will be blown away :-(

I don't have a solution to this, just wanted share my findings. I dpn't think this is a ccache issue, sounds more like a VS issue.
